File: app/src/main/java/com/openmobl/pttDriver/model/Driver.java
package com.openmobl.pttDriver.model;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.NonNull;

public class Driver implements Record, Parcelable {
    private static final String TAG = Driver.class.getName();

    private int mId;
    private String mName;
    private String mType;
    private String mJson;
    private String mDeviceNameMatch;
    private String mWatchForDeviceName;

    public static final Parcelable.Creator<Driver> CREATOR = new Parcelable.Creator<Driver>() {

        @Override
        public Driver createFromParcel(Parcel parcel) {
            return new Driver(parcel);
        }

        @Override
        public Driver[] newArray(int i) {
            return new Driver[i];
        }
    };

    public Driver(String name, String json) {
        mId = -1;
        mName = name;
        mType = ""; // read from JSON
        mJson = json; // validate JSON
        mDeviceNameMatch = null;
        mWatchForDeviceName = null;
    }
    public Driver(int id, String name, String json) {
        this(name, json);

        mId = id;
    }
    public Driver(String name, String type, String json) {
        this(name, json);

        mType = type;
    }
    public Driver(int id, String name, String type, String json) {
        this(id, name, json);

        mType = type;
    }
    public Driver(String name, String type, String json, String deviceNameMatch, String watchForDeviceName) {
        this(name, json);

        mType = type;
        mDeviceNameMatch = deviceNameMatch;
        mWatchForDeviceName = watchForDeviceName;
    }
    public Driver(int id, String name, String type, String json, String deviceNameMatch, String watchForDeviceName) {
        this(id, name, json);

        mType = type;
        mDeviceNameMatch = deviceNameMatch;
        mWatchForDeviceName = watchForDeviceName;
    }

    private Driver(Parcel parcel) {
        readFromParcel(parcel);
    }

    public void setId(int id) { mId = id; }

    @Override
    public int getId() { return mId; }
    @Override
    public String getName() { return mName; }
    public String getType() { return mType; }
    public String getJson() { return mJson; }
    public String getDeviceNameMatch() { return mDeviceNameMatch; }
    public String getWatchForDeviceName() { return mWatchForDeviceName; }
    @Override
    public String getDetails() {
        return "";
    }

    @Override
    public RecordType getRecordType() {
        return RecordType.DRIVER;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int flags) {
        parcel.writeInt(mId);
        parcel.writeString(mName);
        parcel.writeString(mType);
        parcel.writeString(mJson);
        parcel.writeString(mDeviceNameMatch);
        parcel.writeString(mWatchForDeviceName);
    }

    private void readFromParcel(Parcel parcel) {
        mId = parcel.readInt();
        mName = parcel.readString();
        mType = parcel.readString();
        mJson = parcel.readString();
        mDeviceNameMatch = parcel.readString();
        mWatchForDeviceName = parcel.readString();
    }

    public static Driver getEmptyDriver() {
        return new Driver("", "");
    }

    @NonNull
    @Override
    public String toString() {
        return getName();
    }

    public String toStringFull() {
        return "Driver(" + getId() + ", " + getName() + ", " + getType() + ", JSON String)";
    }
}

File: app/src/main/java/com/openmobl/pttDriver/model/Driver.kt
package com.openmobl.pttDriver.model

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import com.openmobl.pttDriver.model.Record.RecordType

class Driver : Record, Parcelable {
    override var id = 0
    override var name: String? = null
        private set
    var type: String? = null
        private set
    var json: String? = null
        private set
    var deviceNameMatch: String? = null
        private set
    var watchForDeviceName: String? = null
        private set

    constructor(name: String?, json: String?) {
        id = -1
        this.name = name
        type = "" // read from JSON
        this.json = json // validate JSON
        deviceNameMatch = null
        watchForDeviceName = null
    }

    constructor(id: Int, name: String?, json: String?) : this(name, json) {
        this.id = id
    }

    constructor(name: String?, type: String?, json: String?) : this(name, json) {
        this.type = type
    }

    constructor(id: Int, name: String?, type: String?, json: String?) : this(id, name, json) {
        this.type = type
    }

    constructor(
        name: String?,
        type: String?,
        json: String?,
        deviceNameMatch: String?,
        watchForDeviceName: String?
    ) : this(name, json) {
        this.type = type
        this.deviceNameMatch = deviceNameMatch
        this.watchForDeviceName = watchForDeviceName
    }

    constructor(
        id: Int,
        name: String?,
        type: String?,
        json: String?,
        deviceNameMatch: String?,
        watchForDeviceName: String?
    ) : this(id, name, json) {
        this.type = type
        this.deviceNameMatch = deviceNameMatch
        this.watchForDeviceName = watchForDeviceName
    }

    private constructor(parcel: Parcel) {
        readFromParcel(parcel)
    }

    override val details: String
        get() = ""
    override val recordType: RecordType
        get() = RecordType.DRIVER

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(id)
        parcel.writeString(name)
        parcel.writeString(type)
        parcel.writeString(json)
        parcel.writeString(deviceNameMatch)
        parcel.writeString(watchForDeviceName)
    }

    private fun readFromParcel(parcel: Parcel) {
        id = parcel.readInt()
        name = parcel.readString()
        type = parcel.readString()
        json = parcel.readString()
        deviceNameMatch = parcel.readString()
        watchForDeviceName = parcel.readString()
    }

    override fun toString(): String {
        return name!!
    }

    fun toStringFull(): String {
        return "Driver(" + id + ", " + name + ", " + type + ", JSON String)"
    }

    companion object {
        private val TAG = Driver::class.java.name
        @JvmField
        val CREATOR: Creator<Driver> = object : Creator<Driver> {
            override fun createFromParcel(parcel: Parcel): Driver {
                return Driver(parcel)
            }

            override fun newArray(i: Int): Array<Driver?> {
                return arrayOfNulls(i)
            }
        }
        val emptyDriver: Driver
            get() = Driver("", "")
    }
}